homeassistant/requirements.py

```diff
@@ -97,12 +97,28 @@ async def async_get_integration_with_requirements(
             deps_to_check.append(check_domain)
 
     if deps_to_check:
-        await asyncio.gather(
+        results = await asyncio.gather(
             *[
                 async_get_integration_with_requirements(hass, dep, done)
                 for dep in deps_to_check
-            ]
+            ],
+            return_exceptions=True,
         )
+        filtered_results = list(
+            filter(
+                lambda result: (
+                    isinstance(result, IntegrationNotFound)
+                    and result.domain not in integration.after_dependencies
+                    or (
+                        isinstance(result, BaseException)
+                        and not isinstance(result, IntegrationNotFound)
+                    )
+                ),
+                results,
+            )
+        )
+        if filtered_results:
+            raise filtered_results[0]
```

Comment:
```suggestion
        for result in results:
            if not isinstance(result, BaseException):
               continue
            if not isinstance(result, IntegrationNotFound) or result.domain not in integration.after_dependencies:
               raise result
```

Reply:
Suggested change to reduce this a bit.

Reply:
I literally just removed that because order matters and my test had to test asyncio.gather's order. By using the filter order doesn't matter anymore.

Reply:
For example, if order matters, then you need to test if an after_dependency happens first and doesn't incorrectly raise an exception. With the filter, that isn't in the list of results at all.

Reply:
I tried the suggested code and the test still passes.

Reply:
It does, but the point is the other implementation I just removed has a potential error case that the new code eliminates.

Reply:
Ok, I guess it's fine.